### PR TITLE
Summation for custom score only restraints, and fixing path issues

### LINKED ISSUE #2: Reding generic restraints
Allow user to read a generic restraint by giving the stat file field name.

### LINKED ISSUE #3: analysis_dir does not get created as a new directory
lines like  (line 1317) fig.savefig(self.analysis_dir+file_out) and others just add the analysis_dir as a filename prefix to the output files so that the intended purpose i.e., have all the analysis output in a separate <analysis_dir> directory is not carried out.

### PR DESCRIPTION
**Pull request by tanmoy7989**
Cosmetic changes
-------------------------
1) Fixed file paths throughout the code. It was an inconsistent mix of ```os.path.join``` and using the backlash and "+" to add paths together. Moving ahead, lets not use the backlash and "+" option; it can break things more often than not.

2) Cleared up some bugs that had appeared in  ```do_extract_models_single_rmf```, since @saltzberg 's last commit. Also tidied up that method, e.g. the ```subprocess.check_call``` was implemented with ```shell=True``` which is a known security flaw and is not recommended; removed it and used the much safer (and simpler) ```shutil.use```.

3) Added an option for figure filename extensions other than 'pdf'. The default option is still pdf.

4)  Added an option ```burn_in``` which is set to a default of 1000. This assumes that the first 1000 frames constitute the burn_in phase of the monte carlo and discards them prior to collecting statistics. This was already achieved through the ```th``` attribute, but it is made explicit now. 

5) Added appropriate documentation to the ```AnalysisTrajectories``` constructor and to the README.

Non-cosmetic changes
-------------------------------
1) Added an option ```detect_equilibraton``` (True by default). Turning this off, turns off the statistical inefficiency based equilibration phase determination, and prevents errors during testing time as reported in [this issue](https://github.com/salilab/imp_analysis_tutorial/issues/4) . Basically, I would want to run a super short, potentially un-converged sampling, and still be able to use the analysis pipeline to test that everything is working well; currently, the equilibration detection routine would throw away all of the frames for such a short trajectory and the rest of the pipeline would collapse. 

3) Added an option ```do_sum``` (True by default) to the ```set_analyze_score_only_restraint``` method. By default, this will add all scores for custom restraints of the same type (i.e. whose restraint handle string starts with the same substring in the stat file). If you want to cluster by individual values of score-only-restraints, just turn this option off.

### CODE REVIEW

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Is it a good idea to have constant defaults for burn_in and nskip? Wouldn't a percentage of the total number of frames make more sense (or no default at all) ?

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Pretty sure "parallely" isn't a word ;)

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: This won't work in Python 2 because `exist_ok` doesn't, er, exist there. This is likely why the code was written this way originally.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: `stat_files = sorted(glob.glob(...))` is cleaner and possibly faster than in-place `stat_files.sort()`.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Perhaps out of scope of this patch, but in general avoid `open(foo).readlines()`. Write as

    with open(foo) as fh:
        for line in fh.readlines():

so that `fh` is definitely closed at the end of the loop (may not be otherwise in Python implementations that don't  use refcounting for garbage collection).

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: flake8 will complain that you don't have spaces around '>' here (and probably elsewhere too).

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Same deal with `sort` vs. `sorted` here.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: I know this is in the original code too... but a bare `except` is rarely what you want. List the specific exceptions you want to catch instead so you don't skip everything on a `SyntaxError` for example.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: This should not be the case... since it looks like below you are doing exactly what `rmf_cat` does anyway. If it is slower, this is a bug and should be fixed.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: I'm sure you know this already but `os.system` is not the right thing to do here. Use `subprocess.check_call` instead and get rid of `rmfcat_string`, e.g:

    subprocess.check_call(['rmf_cat'] + filenames + [output_rmf])

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Oh no, never do this. `shutil.copyfileobj` is probably your friend here, e.g. (untested):

    with open(output_score_file, 'wb') as outf:
        for scorefile in scorefiles:
            with open(scorefile, 'rb') as scoref:
                 shutil.copyfileobj(scoref, outf)

**Review by benmwebb — CHANGES_REQUESTED** [review]
Looks mostly OK to me, although I'll defer to @iecheverria and @saltzberg on whether it makes statistical/scientific sense (please do your reviews, guys). Biggest issue: no test cases. But to be fair, there were none to start with so that would be a big ask.

File: pyext/src/analysis_trajectories.py
Comment (saltzberg) [reply pyext/src/analysis_trajectories.py]: Agree that a percentage is more appropriate for the burn-in default, e.g., 0.05 or 0.1. 

For nskip, I think number of frames makes more sense, but just my personal preference.

File: pyext/src/analysis_trajectories.py
Comment (saltzberg) [inline pyext/src/analysis_trajectories.py:96]: I'd add a catch here for available matplotlib plot formats, otherwise it won't throw an error until after all of the calculations.
```
fig = pl.figure()
available_plot_formats = list(fig.canvas.get_supported_filetypes().keys())
if plot_fmt not in available_plot_formats:
    raise Exception("plot_fmt must be a supported matplotlib file type:", available_plot_formats(
```

**Review by saltzberg — CHANGES_REQUESTED** [review]
Useful additions to PMI analysis and thanks for cleaning up the path issues, restricted names and README.

File: pyext/src/analysis_trajectories.py
Comment (iecheverria) [reply pyext/src/analysis_trajectories.py]: I agree that percentage is appropriate for burn_in, but has to be the number nskip an integer number of frames.

File: pyext/src/analysis_trajectories.py
Comment (iecheverria) [inline pyext/src/analysis_trajectories.py:77]: 200 seems too big. I would suggest 20

File: pyext/src/analysis_trajectories.py
Comment (iecheverria) [reply pyext/src/analysis_trajectories.py:96]: Good catch

File: pyext/src/analysis_trajectories.py
Comment (iecheverria) [inline pyext/src/analysis_trajectories.py]: Remove the word rigorous

File: pyext/src/analysis_trajectories.py
Comment (iecheverria) [inline pyext/src/analysis_trajectories.py]: Beware that this plot is set to png because otherwise (i.e. if it is a pdf or other vector file) it is too big. An alternative is to plot only a fraction of the points.

**Review by iecheverria — CHANGES_REQUESTED** [review]

**Review by iecheverria — APPROVED** [review]

**Review by iecheverria — CHANGES_REQUESTED** [review]
Agree with Ben's comments

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: ```burn_in``` now converted to ```burn_in_fraction``` with a default value of 0.02 (i.e. 2%) of the trajectory length.  ```nskip``` retained as an integer in commit 871884ee953b0ce282d98cb31373d20bc0b9218e

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: fixed.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: changed back to the python 2 compatible form that first runs a ```os.path.isdir``` check.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: using ```sorted()``` now instead of the in-place ```.sort()```

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: all file opening everywhere in ```analysis_trajectories``` now follow this pattern.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: fixed.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: fixed.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: removed bare ```except``` s in ```analysis_trajectories.py``` as well ```equilibration.py```

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: agreed. In commit 871884ee953b0ce282d98cb31373d20bc0b9218e, if ```rmf_cat``` is not found through the ```shutil.which``` search, it raises a ```RuntimeError```

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: done.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py:96]: checking for available plot formats now added in commit 871884ee953b0ce282d98cb31373d20bc0b9218e

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: done.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: Makes sense. 
"plot_clustering_scores" suffix changed back to ".png" in commit 871884ee953b0ce282d98cb31373d20bc0b9218e

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py:77]: ```nskip``` default changed to 20.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: implemented this.

**Comment by benmwebb** [comment]
@tanmoy7989, are you done? If so, probably easiest to just re-request the reviews so we can all take another quick look.

**Comment by tanmoy7989** [comment]
just did.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: IOError is an odd error to raise here. I'd expect something like KeyError normally.

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: 1 decimal place here means that it'll say "0.0" for the default, which looks odd!

File: pyext/src/analysis_trajectories.py
Comment (benmwebb) [inline pyext/src/analysis_trajectories.py]: Same issue here as before with Python 2.

**Review by benmwebb — APPROVED** [review]
Looks mostly OK to me. A few very minor issues, but we can fix these after the merge.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: fixed.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: fixed. Hopefully no more outstanding ```os.makedirs```.

File: pyext/src/analysis_trajectories.py
Comment (tanmoy7989) [reply pyext/src/analysis_trajectories.py]: changed to ```%1.2f```

**Comment by benmwebb** [comment]
@iecheverria , @saltzberg if you could quickly look over this too we should be able to get this merged.

**Review by saltzberg — APPROVED** [review]

**Comment by benmwebb** [comment]
I'm going to go ahead and merge this. We can always back it out or modify if @iecheverria has major issues.

**Pull request merged**
